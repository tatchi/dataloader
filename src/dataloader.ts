export let makeLoader = <T>(batch: (keys: number[]) => Promise<T[]>) => {
	let _pending: [number, (value: T) => void][] = [];
	let isRunnig = false;

	const reset = () => {
		_pending = [];
		isRunnig = false;
	};

	let run = () => {
		isRunnig = true;
		setTimeout(async () => {
			let pending = _pending;
			reset();
			let keys = pending.map(([k, _]) => k);
			let results = await batch(keys);

			results.forEach((user, i) => {
				const [_, resolve] = pending[i]!;
				resolve(user);
			});
		}, 0);
	};

	const load = async (k: number): Promise<T> => {
		const promise = new Promise((r) => {
			_pending.push([k, r]);
		});
		if (!isRunnig) {
			run();
		}

		return promise as any;
	};

	return {
		load,
	};
};
