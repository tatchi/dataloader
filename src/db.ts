const wait = (ms: number): Promise<void> =>
	new Promise((resolve) => {
		setTimeout(() => {
			resolve();
		}, ms);
	});

export interface User {
	id: string;
	name: string;
}

export interface Post {
	id: string;
	title: string;
	authorId: string;
	statsId: string;
}

export interface Stats {
	id: string;
	viewCount: number;
}

export const toUserId = (id: number | string) => `user:${id}`;
export const toPostId = (id: number | string) => `post:${id}`;
export const toStatsId = (id: number | string) => `stats:${id}`;

// Sample Users
export const users: User[] = [
	{ id: toUserId(1), name: 'Alice' },
	{ id: toUserId(2), name: 'Bob' },
	{ id: toUserId(3), name: 'Charlie' },
];

const posts: Post[] = [
	// Posts for User 1 (Alice)
	...Array.from({ length: 5 }, (_, i) => ({
		id: toPostId(i + 1),
		title: `Post ${i + 1} by Alice`,
		authorId: 'user:1',
		statsId: toStatsId(i + 1),
	})),
	// Posts for User 2 (Bob)
	...Array.from({ length: 5 }, (_, i) => ({
		id: toPostId(i + 6),
		title: `Post ${i + 1} by Bob`,
		authorId: 'user:2',
		statsId: toStatsId(i + 6),
	})),
	// Posts for User 3 (Charlie)
	...Array.from({ length: 5 }, (_, i) => ({
		id: toPostId(i + 11),
		title: `Post ${i + 1} by Charlie`,
		authorId: 'user:3',
		statsId: toStatsId(i + 11),
	})),
];

// Sample Stats with Corresponding Entries for Each Post
const stats: Stats[] = [
	...Array.from({ length: 15 }, (_, i) => ({
		id: toStatsId(i + 1),
		viewCount: i + 1,
	})),
];

export const makeDb = ({ enableLog }: { enableLog: boolean }) => {
	const log: Console['log'] = (msg) => {
		if (!enableLog) return;
		console.log(msg);
	};

	const getUserById = async (id: string): Promise<User> => {
		log(`getUserById ${id}`);
		await wait(200);
		return users.find((u) => u.id === id)!;
	};

	const getUserByIds = async (ids: string[]): Promise<User[]> => {
		log(`getUserByIds [${ids.join(', ')}]`);
		await wait(200);
		// log(`SELECT * FROM USERS WHERE id in [${ids.join(',')}]`);
		return ids.map((id) => users.find((u) => u.id === id)!);
	};

	const getStatsById = async (id: string): Promise<Stats> => {
		log(`getStatsById ${id}`);
		await wait(200);
		return stats.find((st) => st.id === id)!;
	};

	const getStatsByIds = async (ids: string[]): Promise<Stats[]> => {
		log(`getStatsByIds [${ids.join(', ')}]`);
		await wait(200);
		return ids.map((id) => stats.find((st) => st.id === id)!);
	};

	const getPosts = async (): Promise<Post[]> => {
		log(`getPosts`);
		await wait(200);

		return posts;
	};

	const getFunFact = async (n: number): Promise<string> => {
		wait(200);
		return `Funfatch for viewcount = ${n}`;
	};

	const getFunFacts = async (n: number[]): Promise<string[]> => {
		wait(200);
		return n.map((n) => `Funfatch for viewcount = ${n}`);
	};

	return {
		getUserById,
		getUserByIds,
		getStatsById,
		getPosts,
		getFunFact,
		getStatsByIds,
		getFunFacts,
	};
};
