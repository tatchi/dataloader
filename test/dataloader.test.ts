import { expect, test, describe } from 'bun:test';
import { makeLoader } from '~/dataloader';
import { getUserByIds } from '~/db';

describe('dataloader', () => {
	test('getUserByIds', async () => {
		const userLoader = makeLoader((ids) => getUserByIds(ids));

		const ids = [1, 2, 3, 4, 5];

		const result = await Promise.all(ids.map((id) => userLoader.load(id)));

		expect(result).toEqual(
			ids.map((id) => ({
				userId: id,
			}))
		);
	});
});
