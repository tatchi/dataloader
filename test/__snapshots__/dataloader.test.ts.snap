// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`dataloader getUserByIds 1`] = `
[
  "getUserByIds [user:1, user:2, user:3, user:4, user:5]",
]
`;

exports[`dataloader buildPostsSummary serial 1`] = `
[
  "getPosts",
  "getUserById user:1",
  "getStatsById stats:1",
  "getFunFact 1",
  "getUserById user:1",
  "getStatsById stats:2",
  "getFunFact 2",
  "getUserById user:1",
  "getStatsById stats:3",
  "getFunFact 3",
  "getUserById user:1",
  "getStatsById stats:4",
  "getFunFact 4",
  "getUserById user:1",
  "getStatsById stats:5",
  "getFunFact 5",
  "getUserById user:2",
  "getStatsById stats:6",
  "getFunFact 6",
  "getUserById user:2",
  "getStatsById stats:7",
  "getFunFact 7",
  "getUserById user:2",
  "getStatsById stats:8",
  "getFunFact 8",
  "getUserById user:2",
  "getStatsById stats:9",
  "getFunFact 9",
  "getUserById user:2",
  "getStatsById stats:10",
  "getFunFact 10",
  "getUserById user:3",
  "getStatsById stats:11",
  "getFunFact 11",
  "getUserById user:3",
  "getStatsById stats:12",
  "getFunFact 12",
  "getUserById user:3",
  "getStatsById stats:13",
  "getFunFact 13",
  "getUserById user:3",
  "getStatsById stats:14",
  "getFunFact 14",
  "getUserById user:3",
  "getStatsById stats:15",
  "getFunFact 15",
]
`;

exports[`dataloader buildPostsSummary parallel 1`] = `
[
  "getPosts",
  "getUserById user:1",
  "getStatsById stats:1",
  "getUserById user:1",
  "getStatsById stats:2",
  "getUserById user:1",
  "getStatsById stats:3",
  "getUserById user:1",
  "getStatsById stats:4",
  "getUserById user:1",
  "getStatsById stats:5",
  "getUserById user:2",
  "getStatsById stats:6",
  "getUserById user:2",
  "getStatsById stats:7",
  "getUserById user:2",
  "getStatsById stats:8",
  "getUserById user:2",
  "getStatsById stats:9",
  "getUserById user:2",
  "getStatsById stats:10",
  "getUserById user:3",
  "getStatsById stats:11",
  "getUserById user:3",
  "getStatsById stats:12",
  "getUserById user:3",
  "getStatsById stats:13",
  "getUserById user:3",
  "getStatsById stats:14",
  "getUserById user:3",
  "getStatsById stats:15",
  "getFunFact 1",
  "getFunFact 2",
  "getFunFact 3",
  "getFunFact 4",
  "getFunFact 5",
  "getFunFact 6",
  "getFunFact 7",
  "getFunFact 8",
  "getFunFact 9",
  "getFunFact 10",
  "getFunFact 11",
  "getFunFact 12",
  "getFunFact 13",
  "getFunFact 14",
  "getFunFact 15",
]
`;

exports[`dataloader buildPostsSummary with own datalader (no cache) 1`] = `
[
  "getPosts",
  "getUserByIds [user:1, user:1, user:1, user:1, user:1, user:2, user:2, user:2, user:2, user:2, user:3, user:3, user:3, user:3, user:3]",
  "getStatsByIds [stats:1, stats:1, stats:2, stats:2, stats:3, stats:3, stats:4, stats:4, stats:5, stats:5, stats:6, stats:6, stats:7, stats:7, stats:8, stats:8, stats:9, stats:9, stats:10, stats:10, stats:11, stats:11, stats:12, stats:12, stats:13, stats:13, stats:14, stats:14, stats:15, stats:15]",
  "getFunFacts [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]",
]
`;

exports[`dataloader buildPostsSummary datalader lib (cache) 1`] = `
[
  "getPosts",
  "getUserByIds [user:1, user:2, user:3]",
  "getStatsByIds [stats:1, stats:2, stats:3, stats:4, stats:5, stats:6, stats:7, stats:8, stats:9, stats:10, stats:11, stats:12, stats:13, stats:14, stats:15]",
  "getFunFacts [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]",
]
`;
